test/beforeHeadersSend.test.js

```diff
@@ -29,172 +32,241 @@ let setMock;
+                setSpentHosts(url.host, 0);
+                setSpentUrl(url.href, true);
+                setReedemMethod(config.id);
+                let redeemHdrs = beforeSendHeaders(details, url);
+                expect(redeemHdrs.cancel).toBeFalsy();
+                switch (config.id) {
+                    case 1:
+                        expect(redeemHdrs.requestHeaders).toBeFalsy();
+                        break
+                    case 2:
+                        // hCaptcha will always spend on its URLS
+                        expect(redeemHdrs.requestHeaders).toBeTruthy();
+                        break
+                    default:
+                        throw Error(`Unhandled config.id value => ${config.id}`)
+                }
+
+            });
+            test(`redemption method is not reload, config.id = ${config.id}`, () => {
+                setSpendFlag(url.host, true);
+                setSpentHosts(url.host, 0);
+                setSpentUrl(url.href, false);
+                workflow.__set__("REDEEM_METHOD", "invalid");
+                let redeemHdrs = beforeSendHeaders(details, url);
+                expect(redeemHdrs.cancel).toBeFalsy();
+                expect(redeemHdrs.requestHeaders).toBeFalsy();
+            });
+            test(`no token to spend, config.id = ${config.id}`, () => {
+                localStorage = {
+                    "cf-bypass-tokens": `{}`,
+                    "cf-token-count": 0
+                };
+                setSpentUrl(url.href, false);
+                setSpendFlag(url.host, true);
+                let redeemHdrs = beforeSendHeaders(details, url);
+                expect(redeemHdrs.cancel).toBeFalsy();
+                expect(redeemHdrs.requestHeaders).toBeFalsy();
+                expect(getSpendFlag(url.host)).toBeFalsy();
+            });
+        });
+
+        describe("redemption attempted", () => {
+            beforeEach(() => {
+                resetVars();
+                resetSpendVars();
+            });
+            test(`redemption header added (SEND_H2C_PARAMS = false), config.id = ${config.id}`, () => {
+                setSpendFlag(url.host, true);
+                setSpentUrl(url.href, false);
+                setReedemMethod(config.id);
+                workflow.__set__("SEND_H2C_PARAMS", false);
+                let redeemHdrs = beforeSendHeaders(details, url);
+                let reqHeaders = redeemHdrs.requestHeaders;
+                expect(getSpendFlag(url.host)).toBeFalsy();
+                expect(getSpendId([details.requestId])).toBeTruthy();
+                expect(getSpentUrl([url.href])).toBeTruthy();
+                switch (config.id) {
+                    case 1:
+                        expect(getSpentTab([details.tabId]) == url.href).toBeTruthy();
+                        break
+                    case 2:
+                        expect(getSpentTab([details.tabId]) == url.href).toBeFalsy();
```

Comment:
A comment to explain why these are different?

Reply:
They actually shouldn't be, added tab accounting for hCaptcha as well